src/Numerics/Mesh/DSS.jl

```diff
@@ -0,0 +1,273 @@
+using CUDA
+
+"""
+    effect_dss3d!(Q::MPIStateArray,
+        grid::DiscontinuousSpectralElementGrid)
+
+This function computes the 3D direct stiffness summation for all variables in the MPIStateArray.
+
+# Fields
+ - `Q`: MPIStateArray
+ - `grid`: Discontinuous Spectral Element Grid
+"""
+function effect_dss3d!(
+    Q::MPIStateArray,
+    grid::DiscontinuousSpectralElementGrid;
+    max_threads = 256,
+)
+    DA = arraytype(grid)                                      # device array
+    device = arraytype(grid) <: Array ? CPU() : CUDADevice()  # device
+    FT = eltype(Q.data)
+    #----communication--------------------------
+    event = MPIStateArrays.begin_ghost_exchange!(Q)
+    event = MPIStateArrays.end_ghost_exchange!(Q, dependencies = event)
+    wait(event)
+    #----Direct Stiffness Summation-------------
+    vertmap = grid.vertmap
+    edgemap = grid.edgemap
+    facemap = grid.facemap
+
+    vtconn = grid.topology.vtconn
+    fcconn = grid.topology.fcconn
+    edgconn = grid.topology.edgconn
+
+    vtconnoff = grid.topology.vtconnoff
+    edgconnoff = grid.topology.edgconnoff
+
+    nvt, nfc, nedg =
+        length(vtconnoff) - 1, size(fcconn, 1), length(edgconnoff) - 1
+
+    Nq = polynomialorders(grid) .+ 1
+    Nqmax = maximum(Nq)
+    Nemax = Nqmax - 2
+    Nfmax = size(facemap, 1)
+    args = (
+        Q.data,
+        vertmap,
+        edgemap,
+        facemap,
+        vtconn,
+        vtconnoff,
+        edgconn,
+        edgconnoff,
+        fcconn,
+        nvt,
+        nedg,
+        nfc,
+        Nemax,
+        Nfmax,
+    )
+    if device == CPU()
+        effect_dss3d_CPU!(args...)
+    else
+        n_items = nvt + nfc + nedg
+        tx = max(n_items, max_threads)
+        bx = cld(n_items, tx)
+        @cuda threads = (tx) blocks = (bx) effect_dss3d_CUDA!(args...)
+    end
+    return nothing
+end
+
+function effect_dss3d_CUDA!(
+    data::AbstractArray{FT, 3},
+    vertmap::AbstractArray{I, 1},
+    edgemap::AbstractArray{I, 3},
+    facemap::AbstractArray{I, 3},
+    vtconn::AbstractArray{I, 1},
+    vtconnoff::AbstractArray{I, 1},
+    edgconn::AbstractArray{I, 1},
+    edgconnoff::AbstractArray{I, 1},
+    fcconn::AbstractArray{I, 2},
+    nvt::I,
+    nedg::I,
+    nfc::I,
+    Nemax::I,
+    Nfmax::I,
+) where {I <: Int, FT <: AbstractFloat}
+
+    tx = threadIdx().x        # threaid id
+    bx = blockIdx().x         # block id
+    bxdim = blockDim().x      # block dimension
+    glx = tx + (bx - 1) * bxdim # global id
+
+    vx = (glx ≥ 1 && glx ≤ nvt) ? glx : -1
+    ex = (glx > nvt && glx ≤ (nvt + nedg)) ? (glx - nvt) : -1
+    fx =
+        (glx > (nvt + nedg) && glx ≤ (nvt + nedg + nfc)) ? (glx - nvt - nedg) :
+        -1
```

Comment:
Should we break this down into multiple lines? This line break makes me nervous 😬

Reply:
I tried to split, but this is forced by the formatter.

Reply:
How about
```julia
    Lfx = (glx > (nvt + nedg) && glx ≤ (nvt + nedg + nfc))
    fx = Lfx ? (glx - nvt - nedg) : -1
```
?